File: Java/src/storm/xmlbinder/binder/CollectionObjectBinder.java
package storm.xmlbinder.binder;

import org.w3c.dom.Document;
import org.w3c.dom.Node;

import storm.xmlbinder.XmlBinderException;

/**
 * Class to bind xml tag to an object and add it to a collection.
 * 
 * @author Storm <j-storm@hotmail.fr>
 */
public class CollectionObjectBinder extends AbstractCollectionBinder 
{
	/**
	 * The name of the object class to create.
	 */
	protected String m_className = "";
	
	/**
	 * The content binder instance if needed.
	 */
	protected ContentBinder m_contentBinder = null;
	
	/**
	 * Construct an instance without content binder.
	 * Considering that the xml tag is not a text node and have child element.
	 * @param _fieldName : the name of the collection field.
	 * @param _className : the name of the class to create object.
	 */
	public CollectionObjectBinder(String _fieldName, String _className) 
	{
		super(_fieldName);
		this.m_className = _className;
	}
	
	/**
	 * Construct an instance with a content binder.
	 * Consider that the xml element is a text node.
	 * @param _fieldName : the name of the collection field.
	 * @param _className : the name of the class to create object.
	 * @param _contentBinder : the content binder to use.
	 */
	public CollectionObjectBinder(String _fieldName, String _className, ContentBinder _contentBinder)
	{
		this(_fieldName, _className);
		this.m_contentBinder = _contentBinder;
	}
	
	/*
	 * (non-Javadoc)
	 * @see storm.xmlbinder.binder.AbstractBinder#read(java.lang.Object, java.lang.String)
	 */
	@Override
	public Object read(Object _data, String _content) throws XmlBinderException 
	{
		Object newData = createObject(this.m_className);
		add(newData, _data, this.m_fieldName);
		
		if(this.m_contentBinder != null)
		{
			m_contentBinder.read(newData, _content);
		}
		
		return newData;
	}

	/*
	 * (non-Javadoc)
	 * @see storm.xmlbinder.binder.AbstractBinder#write(java.lang.Object, org.w3c.dom.Node, org.w3c.dom.Document)
	 */
	@Override
	public void write(Object _data, Node _node, Document _document) throws XmlBinderException 
	{
		ObjectBinder obj = new ObjectBinder(this.m_fieldName, this.m_className, this.m_contentBinder);
		obj.write(_data, _node, _document);
	}

}

File: C#/Binder/CollectionObjectBinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using XmlBinder.Abstract;
using XmlBinder.Interface;

namespace XmlBinder.Binder
{
	/// <summary>
	/// Binder to add object to IList member.
	/// </summary>
	public class CollectionObjectBinder : AbstractCollectionBinder
	{
		/// <summary>
		/// Type of object to create.
		/// </summary>
		protected Type m_objectType = null;

		/// <summary>
		/// Binder to bind content of xml tag if needed.
		/// </summary>
		protected BinderInterface m_contentBinder = null;

		/// <summary>
		/// Type of object to create.
		/// </summary>
		public Type ClassType
		{
			get
			{
				return m_objectType;
			}
		}

		/// <summary>
		/// Binder to bind xml tag content.
		/// </summary>
		public BinderInterface ContentBinder
		{
			get
			{
				return m_contentBinder;
			}
		}

		/// <summary>
		/// Create a new collection object binder.
		/// </summary>
		/// <param name="_fieldName">The name of the collection field.</param>
		/// <param name="_newType">The type of the object to create.</param>
		/// <param name="_contentBinder">The binder to use to xml content if needed.</param>
		public CollectionObjectBinder(string _fieldName, Type _newType, BinderInterface _contentBinder = null)
			: base(_fieldName)
		{
			m_objectType = _newType;
			m_contentBinder = _contentBinder;
		}

		public override object read(object _source, string _content)
		{
			object data = Activator.CreateInstance(ClassType);
			addElement(_source, FieldName, data);
			if (ContentBinder != null)
			{
				ContentBinder.read(data, _content);
			}
			return data;
		}

		public override string write(object _source)
		{
			if (ContentBinder != null)
			{
				//return ContentBinder.write(_source);
				return ContentBinder.write(ContentBinder.getAffectedField(_source));
			}
			return "";
		}
	}
}
